deblur/deblurring.py

```diff
@@ -0,0 +1,167 @@
+    the hamming distance taken into account, i.e. if the length if `error_dist`
+    is 10, sequences up to 10 - 1 = 9 hamming distance will be taken into
+    account
+    """
+    # Get the sequences
+    seqs = get_sequences(input_seqs)
+
+    # If mean error is not provided, use the same value as read_error
+    mean_error = mean_error if mean_error else read_error
+
+    # if error_list not supplied, use the default (22 mock mixture setup)
+    mod_factor = pow((1 - mean_error), seqs[0].unaligned_length)
+    if not error_dist:
+        error_dist = np.array(
+            [1.0 / mod_factor, pow(read_error, 1) / mod_factor, 0.01, 0.01,
+             0.01, 0.005, 0.005, 0.005, 0.005, 0.005, 0.005, 0.001, 0.001,
+             0.001, 0.001])
+    else:
+        error_dist = np.array(error_dist) / mod_factor
+
+    max_h_dist = len(error_dist)-1
+
+    for seq_i in seqs:
+        # no need to remove neighbors if freq. is <=0
+        if seq_i.frequency <= 0:
+            continue
+
+        # Correct for the fact that many reads are expected to be mutated
+        num_err = error_dist * seq_i.frequency
+        num_err = [val * seq_i.frequency for val in error_dist]
+
+        # if it's low level, just continue
+        if num_err[1] < 0.1:
+            continue
+
+        # Compare to all other sequences and calculate hamming dist
+        seq_i_len = len(seq_i.sequence.rstrip('-'))
+        for seq_j in seqs:
+            # Ignore current sequence
+            if seq_i == seq_j:
+                continue
+
+            # Calculate the hamming distance
+            h_dist = np.count_nonzero(np.not_equal(seq_i.np_sequence,
+                                                   seq_j.np_sequence))
+
+            # If far away, don't need to correct
+            if h_dist > max_h_dist:
+                continue
+
+            # Close, so lets calculate exact distance
+            num_substitutions = 0
+            num_indels = 0
+
+            # We stop checking in the shortest sequence after removing trailing
+            # indels. We need to do this in order to avoid double counting
+            # the insertions/deletions
+            l = min(seq_i_len, len(seq_j.sequence.rstrip('-')))
+            for bi, bj in zip(seq_i.np_sequence[:l], seq_i.np_sequence[:l]):
+                if bi != bj:
+                    if bi == 4 or bj == 4:
+                        num_indels += 1
+                    else:
+                        num_substitutions += 1
+
+            correction_value = num_err[num_substitutions]
+
+            if num_indels > indel_max:
+                correction_value = 0
+            elif num_indels > 0:
+                # remove errors due to (PCR?) indels (saw in 22 mock mixture)
+                correction_value = correction_value * indel_prob
+
+            # met all the criteria - so correct the frequency of the neighbor
+            seq_j.frequency -= correction_value
+
+    result = [s for s in seqs if s.frequency > 0]
```

Comment:
this loop is effectively done twice, why not have `get_sequences` filter out sequences whose frequency is `<= 0`?